### ISSUE TITLE
Is It Normal for indices in compute_synchronisation to Be Hardcoded?

### ISSUE THREAD

**Issue by XiudingCai** [SakanaAI/continuous-thought-machines#6]
Hi,

I have noticed a potential issue in the `compute_synchronisation` function. Specifically, in the following section:

```python
elif self.neuron_select_type == 'random-pairing':
    # For random-pairing, we compute the sync between specific pairs of neurons
    left = activated_state[:, neuron_indices_left]
    right = activated_state[:, neuron_indices_right]
    pairwise_product = left * right
```

It seems that the indices (`neuron_indices_left` and `neuron_indices_right`) are hardcoded rather than being generated randomly for each training iteration. This approach may lead to the same pairs of neurons being selected every time, which contradicts the concept of "random-pairing."

Is this behavior intentional?

Should the neuron indices be re-sampled randomly for each training iteration instead?

Looking forward to your clarification.

Thank you!

**Comment by lukedarlow** [comment]
Hi,

No, this is not an issue. For your reference, Appendix B.2 in the paper discusses the three types of pairings. In the paper we call them 'dense', 'semi-dense', and 'random pairing', these correspond to 'first-last', 'random', and 'random-pairing' in the code. 

It would simply not work to randomly pair neurons per training iterations. There is no conceivable benefit to doing so and it would make it almost impossible to learn. 

The sub-sampling strategies, including random-pairing, are designed to select a smaller representation from the substantially large synchronization representation. For D=1000, for instance, there are over 0.5 million unique neuron-pairings that form the synchronization matrix. Projecting from such a large space to an attention query or logits is needlessly costly, hence the neuron selections.
